src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,445 @@
+            var constantsMapper = analysisContext.Node.ConstantsRemapper;
+            argumentTypeName = analysisContext.Node.ArgumentTypeName;
+            nodesRemapping.Add(builderNode, rewrittenBuildersNode);
+            constantMappers.Add(constantsMapper);
+        }
+
+        var rewrittenSyntaxNode = syntaxNode.ReplaceNodes(nodesRemapping.Keys, (n, _) => nodesRemapping[n]);
+
+        if (IsCompoundExpression(syntaxKind))
+        {
+            var rewrittenAssignmentExpression = rewrittenSyntaxNode as AssignmentExpressionSyntax;
+            var binaryExpressionSyntaxKind = GenerateBinaryExpressionSyntaxKind(syntaxKind);
+            rewrittenSyntaxNode = SyntaxFactory.BinaryExpression(binaryExpressionSyntaxKind, rewrittenAssignmentExpression.Left, rewrittenAssignmentExpression.Right);
+        }
+
+        var rewrittenBuildersExpression = rewrittenSyntaxNode switch
+        {
+            AssignmentExpressionSyntax @assignmentExpression => @assignmentExpression.Right,
+            VariableDeclaratorSyntax @variableDeclarator => @variableDeclarator.Initializer.Value,
+            _ => rewrittenSyntaxNode
+        };
+
+        var (finalBuildersExpression, finalConstantMapper) = ConstantsMapperComposer.Compose(rewrittenBuildersExpression, constantMappers);
+        return new ExpressionAnalysisContext(new ExpressionAnalysisNode(rewrittenBuildersExpression, argumentTypeName, finalBuildersExpression, finalConstantMapper, new Location[] { }));
+    }
+
+    private static bool ProcessDeclaration(VariableDeclaratorSyntax variableDeclarator, ContextInformation contextInformation, HashSet<string> dirtyVariables, bool displayDiagnostics)
+    {
+        var evaluationResult = PreEvaluate(variableDeclarator, contextInformation, null, displayDiagnostics);
+        var variableName = variableDeclarator.Identifier.ValueText;
+
+        if (evaluationResult.CanEvaluate)
+        {
+            var variableValue = evaluationResult.SyntaxNodeToVariableValueMap.Values.First();
+            var generateNewContext = evaluationResult.SyntaxNodeToVariableValueMap.Count() > 1;
+
+            if (generateNewContext)
+            {
+                var expressionAnalysisContext = GenerateNewExpressionAnalysisContext(variableDeclarator, evaluationResult.SyntaxNodeToVariableValueMap);
+                variableValue = new VariableValue(expressionAnalysisContext, new List<Location>());
+            }
+
+            dirtyVariables.Add(variableName);
+            contextInformation.VariableValues.Add(variableName, variableValue);
+
+            var originalExpression = variableValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (contextInformation.BuilderNodesProcessed.Add(originalExpression))
+            {
+                contextInformation.VariableValuesList.Add(variableValue);
+            }
+        }
+        else
+        {
+            RemoveVariablesFromContext(new List<string> { variableName }, contextInformation);
```

Comment:
Maybe add an overload RemoveVariablesFromContext(string variableToRemove)
And reuse it from RemoveVariablesFromContext(IEnumerable<string> variablesToRemove)

Reply:
Done